server/src-lib/Hasura/Server/Query.hs

```diff
@@ -301,31 +301,34 @@ queryNeedsReload (RQV2 qi) = case qi of
   RQV2SetTableCustomFields _ -> True
   RQV2TrackFunction _        -> True
 
--- TODO: RQSelect query should also be run in READ ONLY mode.
--- But this could be part of console's bulk statement and hence should be added after console changes
 getQueryAccessMode :: (MonadError QErr m) => RQLQuery -> m Q.TxAccess
-getQueryAccessMode (RQV1 q) =
-  case q of
-    RQRunSql RunSQL{rTxAccessMode} -> pure rTxAccessMode
-    RQBulk qs -> assertAllTxAccess (zip [0::Integer ..] qs)
-    _                              -> pure Q.ReadWrite
+getQueryAccessMode q = (fromMaybe Q.ReadOnly) <$> getQueryAccessMode' q
   where
-    assertAllTxAccess = \case
-      [] -> throw400 BadRequest "expected atleast one query in bulk"
-      (_i, q1):[] -> getQueryAccessMode q1
-      q1:q2:qs -> assertSameTxAccess q1 q2 >> assertAllTxAccess (q2:qs)
-
-    assertSameTxAccess (i1, q1) (i2, q2) = do
-      accessModeQ1 <- getQueryAccessMode q1
-      accessModeQ2 <- getQueryAccessMode q2
-      if (accessModeQ1 /= accessModeQ2)
-        then
-        throw400 BadRequest $ "incompatible access mode requirements in bulk query: "
-        <> "$.args[" <> (T.pack $ show i1) <> "] requires " <> (T.pack $ show accessModeQ1) <> ", "
-        <> "$.args[" <> (T.pack $ show i2) <> "] requires " <> (T.pack $ show accessModeQ2)
-        else
-        pure accessModeQ1
-getQueryAccessMode (RQV2 _) = pure Q.ReadWrite
+    getQueryAccessMode' :: (MonadError QErr m) => RQLQuery -> m (Maybe Q.TxAccess)
+    getQueryAccessMode' (RQV1 q') =
+      case q' of
+        RQSelect _ -> pure Nothing -- RQSelect is don't care i.e. works with both types of TxAccess
+        RQCount _ -> pure Nothing  -- RQCount is don't care i.e. works with both types of TxAccess
+        RQRunSql RunSQL{rTxAccessMode} -> pure $ Just rTxAccessMode
+        RQBulk qs -> assertAllAccessMode Nothing (zip [0 ..] qs)
+        _                              -> pure $ Just Q.ReadWrite
+      where
+        assertAllAccessMode :: (MonadError QErr m) => Maybe Q.TxAccess -> [(Integer, RQLQuery)] -> m (Maybe Q.TxAccess)
+        assertAllAccessMode expectedAccessMode = \case
+          [] -> throw400 BadRequest "expected atleast one query in bulk"
+          q1:[] -> assertAccessMode expectedAccessMode q1
+          q1:qs -> assertAccessMode expectedAccessMode q1 >>= (flip assertAllAccessMode qs)
```

Comment:
would foldM be cuter here?

Reply:
Yes, I think the way I would expect to see this written is to have a function that checks compatibility between two access modes, returning their least upper bound and raising an error if they’re incompatible. Something like this:

```haskell
reconcileAccessModes :: Maybe Q.TxAccess -> Maybe Q.TxAccess -> m (Maybe Q.TxAccess)
reconcileAccessModes Nothing mode = pure mode
reconcileAccessModes mode Nothing = pure mode
reconcileAccessModes (Just mode1) (Just mode2)
  | mode1 == mode2 -> pure mode1
  | otherwise -> {- raise error -}
```

Then you could pass that function to `foldM` with `Nothing` as the initial element to fold over the list of required access modes.